### PR TITLE
fix: cpu contention on jwk reads + suppress duplicate jwk generation

### LINKED ISSUE #3863: GetOrGenerateKeys locking leads to significant service degredation with high request saturation
### Preflight checklist

- [ ] I could not find a solution in the existing issues, docs, nor discussions.
- [X] I agree to follow this project's [Code of Conduct](https://github.com/ory/hydra/blob/master/CODE_OF_CONDUCT.md).
- [X] I have read and am following this repository's [Contribution Guidelines](https://github.com/ory/hydra/blob/master/CONTRIBUTING.md).
- [X] I have joined the [Ory Community Slack](https://slack.ory.sh).
- [X] I am signed up to the [Ory Security Patch Newsletter](https://www.ory.sh/l/sign-up-newsletter).

### Ory Network Project

_No response_

### Describe the bug

When operating Hydra in production, it has been observed through tracing that the `GetOrGenerateKeys` is called routinely for the Authorization Endpoint, Token Endpoint, and WellKnown endpoints. That function leverages two RWMutexes to store the keysets in memory.

As identified in https://github.com/ory/hydra/discussions/3662, there can be problems when the tail latency of the database query to retrieve JWKs is above a certain threshold under highly concurrent conditions.

Additionally, when considering the locking logic for `GetOrGenerateKeys`, there may be an opportunity to leverage `RLock()` when reading from the database and then acquiring the write lock only when generating to enable more throughput for this endpoint.

### Reproducing the bug

To recreate the issue locally:

## Strategy

1. **Database Latency** - Run `hydra` with an artificially set latency for the `persister_jwk`'s `GetKeySet` that mimics real-world conditions (>= 25ms).
2. **Request Saturation** - Using any load testing tool like [bombardier](https://github.com/codesenberg/bombardier) to saturate the server with hundreds of concurrent client connections against `/.well-known/openid-configuration` endpoint.

## Steps

0. Install bombardier: `go install github.com/codesenberg/bombardier@latest`
1. Run cockroachdb instance: `docker compose -f quickstart.yml -f quickstart-cockroach.yml up -d cockroachd`
2. Run migrations: `docker compose -f quickstart.yml -f quickstart-cockroach.yml run hydra-migrate`
3. Add artificial latency to [`persister_jwk.go`](https://github.com/awill1988/hydra/blob/9cc5f28825bb80398a9830d4fcb220147f74504c/persistence/sql/persister_jwk.go#L170): `time.Sleep(25 * time.Millisecond)`
4. Run code: `DSN='cockroach://root@localhost:26257/defaultdb?sslmode=disable&max_conns=20&max_idle_conns=4' go run . serve all --dev -c contrib/quickstart/5-min/hydra.yml`
5. Run load test with `270` - `300` concurrent connections `bombardier -d 30s -t 30s -a -c 270 -l http://localhost:4444/.well-known/openid-configuration`

## Expected Outcome

Request saturation would result in a moderate service degradation. Here's a test with only 100 concurrent connections:

```
❯ bombardier -d 30s -t 30s -a -c 100 -l http://localhost:4444/.well-known/openid-configuration
Bombarding http://localhost:4444/.well-known/openid-configuration for 30s using 100 connection(s)
[=====================================================================================================================] 30s
Done!
Statistics        Avg      Stdev        Max
  Reqs/sec        26.14      17.31      57.67
  Latency         3.62s   708.85ms      3.90s
  Latency Distribution
     50%      3.83s
     75%      3.86s
     90%      3.89s
     95%      3.89s
     99%      3.90s
  HTTP codes:
    1xx - 0, 2xx - 881, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:    54.73KB/s
```

## Actual Results

If the logarithmic growth in request handler timing reaches 10s, request latency degrades significantly.

```
❯ bombardier -d 30s -t 30s -a -c 270 -l http://localhost:4444/.well-known/openid-configuration
Bombarding http://localhost:4444/.well-known/openid-configuration for 30s using 270 connection(s)
[=====================================================================================================================] 30s
Done!
Statistics        Avg      Stdev        Max
  Reqs/sec         8.70      17.32      52.68
  Latency        28.31s     22.94s      0.85m
  Latency Distribution
     50%     41.07s
     75%      0.85m
     90%      0.85m
     95%      0.85m
     99%      0.85m
  HTTP codes:
    1xx - 0, 2xx - 519, 3xx - 0, 4xx - 0, 5xx - 0
    others - 10
  Errors:
    the server closed connection before returning the first response byte. Make sure the server returns 'Connection: close' response header before closing the connection - 10
  Throughput:    19.69KB/s
```

## Potential Fix

If [`GetOrGenerateKeys`](https://github.com/awill1988/hydra/blob/9cc5f28825bb80398a9830d4fcb220147f74504c/jwk/helper.go#L46) were to leverage a read lock and only acquiring a write lock when it wants to generate, then this request saturation does not have the same effect, but this may not provide the desired synchronization guarantees intended by the authors.

```
 func GetOrGenerateKeys(ctx context.Context, r InternalRegistry, m Manager, set, kid, alg string) (private *jose.JSONWebKey, err error) {
-       getLock(set).Lock()
-       defer getLock(set).Unlock()
-
+       getLock(set).RLock()
        keys, err := m.GetKeySet(ctx, set)
+       getLock(set).RUnlock()
+
        if errors.Is(err, x.ErrNotFound) || keys != nil && len(keys.Keys) == 0 {
                r.Logger().Warnf("JSON Web Key Set \"%s\" does not exist yet, generating new key pair...", set)
+               getLock(set).Lock()
+               defer getLock(set).Unlock()
                keys, err = m.GenerateAndPersistKeySet(ctx, set, kid, alg, "sig")
                if err != nil {
                        return nil, err
```

```
❯ bombardier -d 30s -t 30s -a -c 270 -l http://localhost:4444/.well-known/openid-configuration
Bombarding http://localhost:4444/.well-known/openid-configuration for 30s using 270 connection(s)
[===============================================================================================================================================] 30s
Done!
Statistics        Avg      Stdev        Max
  Reqs/sec      1379.25     412.60    3602.03
  Latency      195.31ms   321.83ms      9.35s
  Latency Distribution
     50%   110.06ms
     75%   199.68ms
     90%   377.27ms
     95%   565.53ms
     99%      1.44s
  HTTP codes:
    1xx - 0, 2xx - 41610, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:     2.82MB/s
```

### Relevant log output

_No response_

### Relevant configuration

_No response_

### Version

v2.2.0

### On which operating system are you observing this issue?

Linux

### In which environment are you deploying?

Kubernetes

### Additional Context

This was observed in a production Kubernetes deployment where 4 replicas of a hydra pods were getting thousands of requests per second and it led to a significant increase in latency. Upon examination of the traces, the large gap between spans before `persistence.sql.GetKeySet` was found to be a signature for the issue.

### PR DESCRIPTION
**Pull request by terev**
Using the reproduction steps included in #3863 I'm able to confirm this fixes the read contention:
```zsh
bombardier -d 30s -t 30s -a -c 270 -l http://localhost:4444/.well-known/openid-configuration
Bombarding http://localhost:4444/.well-known/openid-configuration for 30s using 270 connection(s)
[==============================================================================================================================================================] 30s
Done!
Statistics        Avg      Stdev        Max
  Reqs/sec      3768.36    2589.36   12964.59
  Latency       71.76ms    29.92ms   579.56ms
  Latency Distribution
     50%    67.30ms
     75%    82.45ms
     90%    98.98ms
     95%   113.15ms
     99%   184.51ms
  HTTP codes:
    1xx - 0, 2xx - 112946, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:     7.70MB/s
```

I also noticed the .well-known/jwks.json route generates extra keys (if hammered before initialized) so I changed that handler to utilize the same duplicate suppression code.

## Related issue(s)
Closes #3863 

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [ ] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

### CODE REVIEW

File: jwk/helper.go [hunk 1]
```diff
@@ -12,69 +12,67 @@ import (
 	"crypto/x509"
 	"encoding/json"
 	"encoding/pem"
-	"sync"
+
+	"golang.org/x/sync/singleflight"
```
Comment (awill1988) [inline jwk/helper.go:16]: nice, til

File: jwk/helper_test.go [hunk 2]
```diff
@@ -229,7 +230,7 @@ func TestGetOrGenerateKeys(t *testing.T) {
 	t.Run("Test_Helper/Run_GetOrGenerateKeys_With_GetKeySetError", func(t *testing.T) {
```
Comment (awill1988) [inline jwk/helper_test.go:230]: should test name reflect func name change?

`Run_GetOrGenerateKeys_With_GetKeySetError` -> `Run_GetOrGenerateKeySetPrivateKey_With_GetKeySetError`

Comment (terev) [reply jwk/helper_test.go:230]: Oh yeah good call. Might do a little more refactoring here because the name doesn't feel quite right.

Comment (terev) [reply jwk/helper_test.go]: Changed my mind, left the name the same and fixed the test names.

**Review by aeneasr — APPROVED** [review]

**Comment by aeneasr** [comment]
Great job everyone, thanks!

**Comment by terev** [comment]
🎉 Thanks for the well written initial report @awill1988 !

**Comment by awill1988** [comment]
Awesome fix @terev!

**Comment by aeneasr** [comment]
Unfortunately we need to revert this as it is messing with a downstream implementation in Ory Network. I think the code is still not working as fully intended.

**Pull request merged**